### ISSUE TITLE
View get seems slow (or Am I messing it up?)

### ISSUE THREAD

**Issue by Paolo5150** [skypjack/entt#822]
Hi future reader,

I just got entt, and I really like the simplicity of it.

I have my quick test, small rendering engine running using OpenGL, I noticed that retrieving components with "view.get" is slowing me down, quite significantly (like nearly halving FPS).

Quick screenshot from performance profiler:
![image](https://user-images.githubusercontent.com/8026573/147849783-bd490239-fddb-4d7b-93ad-b319b83085c3.png)

I have my entities with a Transform and TileRenderer components assigned, both very simple structs.

Is that expected? Am I doing something wrong?

**Comment by skypjack** [comment]
Not sure what the image is supposed to represent, so I'll try to answer you more generally.

It's unlikely that relying on the `get` actually halves access times. I'm lucky enough to work on some very large projects that make extensive use of it and fortunately times are different. I would be in serious troubles otherwise. 😅
The first idea that comes to my mind is that you're measuring it in debug? Some platforms (_cough coff_ i.e. Windows _coff cough_) have a huge overhead in this case, this could explain your measure (even though it still seems way larger than expected).
I don't know what to suggest otherwise. It's difficult to investigate without a minimum and reproducible example.

In any case, it's also true that using the `get` is slightly slower than relying on the optimized iteration methods already offered by a view, such as the `.each` member functions:
```cpp
// with a callback
view.each([](auto entity, auto &transform, auto &tile_renderer) {
    // ...
});

// with an extended iterator
for(auto [entity, transform, tile_renderer]: view.each()) {
    // ...
}
```
In both cases, `EnTT` does skip a lot of instructions and can optimize the whole thing quite a lot under the hood.
As a rule of thumb, the `get` is good when you want to retrieve elements out of order or when you don't want to get all components in a view.

**Comment by Paolo5150** [comment]
Thanks for the quick answer!

The image is the visual studio performance profiler, it's meant to show that most of the time spent is on that line, where the view is used.

Following your suggestion, yes, it seems to be performing better. I'll keep profiling the code, but for now this can probably be closed, your solution has already helped, thanks again!

**Comment by collinalexbell** [comment]
I'd like to add my experience with view.get. I was using it ~12 times per frame to fetch all my Renderable positions. My program would run fine for a while and then it would randomly start using 100% CPU only to go back to running fine again after 30 seconds or so. I traced the CPU consumption down to `memmove_avx_unaligned_erms` but I can't see what is calling it. After replacing the view.get with iteration over view.each, the problem seems to have disappeared.

Here is the code diff if anyone is interested:

```
-  if(!lightView.empty()) {
-    auto lightEntity = lightView.front();
-    auto lightPositionable = modelView.get<Positionable>(lightEntity);
-    auto light = lightView.get<Light>(lightEntity);
-    shader->setVec3("lightPos", lightPositionable.pos);
+  shader->setVec3("lightPos", glm::vec3(0, 0, 0));
+  shader->setVec3("lightColor", glm::vec3(0.5, 0.5, 0.5));
+
+  entt::entity lightEntity;
+  bool hasLight = false;
+
+  for(auto [entity, light, positionable]: lightView.each()) {
+    shader->setVec3("lightPos", positionable.pos);
     shader->setVec3("lightColor", light.color);
-  } else {
-    shader->setVec3("lightPos", glm::vec3(0,0,0));
-    shader->setVec3("lightColor", glm::vec3(0.5,0.5,0.5));
+    lightEntity = entity;
+    hasLight = true;
   }
-    
 
-  for(auto entity: modelView) {
-    auto m = modelView.get<Model>(entity);
-    auto p = modelView.get<Positionable>(entity);
-    if(registry->all_of<Light>(entity)) {
+  for(auto [entity, p, m]: modelView.each()) {
+    if(hasLight && lightEntity == entity) {
       shader->setBool("isLight", true);
     }
```

**Comment by skypjack** [comment]
Uhm no @collinalexbell getting an element doesn't rearrange the storage in any case. I don't think that's the cause.
Invoking `get` again and again is less prone to inlining though, although I don't know why this should emit a `memmove_avx_unaligned_erms`. Again, it seems odd to me.
Anyway, 12 times is 🤷‍♂️ irrelevant? I work on a very large project where using `get` hundreds if not thousands of times in a tick is common and we aren't observing the same issue with the CPU. So your problem is curious, that's for sure. Maybe it was due to something else being done at the same time? Hard to say, especially for me.
Something you should avoid in your code is this btw:
```cpp
if(registry->all_of<Light>(entity)) {
```
Long story short, querying the registry already boils down to looking up the right pool to which to forward your request.
This is much better instead since it gets rid of all lookups at once (because views _know_ 😅):
```cpp
auto lightView = registry.view<Light>();
for(auto entity: modelView) {
    // ...
    if(view.contains(entity)) { /* ... */ }
    // ...
}
```
